beater/server.go

```diff
@@ -225,7 +195,59 @@ func newGRPCServer(
 	return srv, nil
 }
 
-func (s server) run() error {
+func (s *server) configure(
+	logger *logp.Logger,
+	info beat.Info,
+	cfg *config.Config,
+	tracer *apm.Tracer,
+	reporter publish.Reporter,
+	sourcemapStore *sourcemap.Store,
+	batchProcessor model.BatchProcessor,
+) error {
+	s.agentcfgFetchReporter = agentcfg.NewReporter(agentcfg.NewFetcher(cfg), batchProcessor, 30*time.Second)
+	ratelimitStore, err := ratelimit.NewStore(
+		cfg.AgentAuth.Anonymous.RateLimit.IPLimit,
+		cfg.AgentAuth.Anonymous.RateLimit.EventLimit,
+		3, // burst multiplier
+	)
+	if err != nil {
+		return err
+	}
+	if s.agentCancel != nil {
+		s.agentCancel()
+		s.startFetchReporter(s.agentCtx)
+	}
+	return s.httpServer.configure(
+		logger,
+		info,
+		cfg,
+		tracer,
+		reporter,
+		batchProcessor,
+		s.agentcfgFetchReporter,
+		ratelimitStore,
+		sourcemapStore,
+	)
+}
+
+func (s *server) startFetchReporter(ctx context.Context) {
+	s.agentCtx = ctx
+	ctx, s.agentCancel = context.WithCancel(s.agentCtx)
+	go s.agentcfgFetchReporter.Run(ctx)
+}
```

Comment:
The old fetchReporter needs to be stopped and the new one started when the config is updated because of its internal run loop that batches and sends config updates.